File: android-test/app/src/test/java/co/elastic/apm/android/test/features/centralconfig/fetcher/CentralConfigurationFetcherTest.java
package co.elastic.apm.android.test.features.centralconfig.fetcher;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import androidx.annotation.NonNull;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

import co.elastic.apm.android.sdk.connectivity.auth.AuthConfiguration;
import co.elastic.apm.android.sdk.internal.configuration.impl.ConnectivityConfiguration;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.CentralConfigurationFetcher;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.ConfigurationFileProvider;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.FetchResult;
import co.elastic.apm.android.sdk.internal.services.preferences.PreferencesService;
import co.elastic.apm.android.test.BuildConfig;
import co.elastic.apm.android.test.testutils.base.BaseRobolectricTest;
import okhttp3.HttpUrl;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

@SuppressWarnings("ConstantConditions")
public class CentralConfigurationFetcherTest extends BaseRobolectricTest implements ConfigurationFileProvider {
    private PreferencesService preferences;
    private ConnectivityConfiguration connectivity;
    private File configurationFile;
    private CentralConfigurationFetcher fetcher;
    private MockWebServer webServer;

    @Rule
    public final TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Before
    public void setUp() throws IOException {
        setUpConnectivity();
        preferences = mock(PreferencesService.class);
        configurationFile = temporaryFolder.newFile("configFile.json");

        fetcher = new CentralConfigurationFetcher(connectivity, this, preferences);
    }

    @After
    public void tearDown() throws IOException {
        webServer.shutdown();
    }

    private void setUpConnectivity() {
        webServer = new MockWebServer();
        connectivity = mock(ConnectivityConfiguration.class);
        setConnectivityEndpoint("");
    }

    private void setConnectivityEndpoint(String path) {
        doReturn("http://" + webServer.getHostName() + ":" + webServer.getPort() + path).when(connectivity).getEndpoint();
    }

    @Test
    public void whenResponseIsOk_respondWithConfigurationHasChanged() throws IOException {
        enqueueSimpleResponse();

        FetchResult fetch = fetcher.fetch();

        assertTrue(fetch.configurationHasChanged);
    }

    @Test
    public void whenResponseIsNotOk_respondWithConfigurationHasNotChanged() throws IOException {
        enqueueResponse(getResponse(304, ""));

        FetchResult fetch = fetcher.fetch();

        assertFalse(fetch.configurationHasChanged);
    }

    @Test
    public void whenPreparingRequest_sendQueryParams() throws IOException, InterruptedException {
        enqueueSimpleResponse();

        fetcher.fetch();

        HttpUrl recordedRequestUrl = webServer.takeRequest().getRequestUrl();
        assertEquals("my-app", recordedRequestUrl.queryParameter("service.name"));
        assertEquals(BuildConfig.BUILD_TYPE, recordedRequestUrl.queryParameter("service.environment"));
    }

    @Test
    public void whenPreparingRequest_keep_original_url_path() throws IOException, InterruptedException {
        enqueueSimpleResponse();

        setConnectivityEndpoint("/some/path");
        fetcher = new CentralConfigurationFetcher(connectivity, this, preferences);
        fetcher.fetch();

        HttpUrl recordedRequestUrl = webServer.takeRequest().getRequestUrl();
        assertEquals("/some/path/config%2Fv1%2Fagents", recordedRequestUrl.encodedPath());
    }

    @Test
    public void whenPreparingRequest_sendContentType() throws IOException, InterruptedException {
        enqueueSimpleResponse();

        fetcher.fetch();

        assertEquals("application/json", webServer.takeRequest().getHeader("Content-Type"));
    }

    @Test
    public void whenPreparingRequest_andThereIsAuthenticationAvailable_sendAuthHeader() throws
            IOException, InterruptedException {
        String authHeaderValue = "Bearer something";
        AuthConfiguration authConfiguration = mock(AuthConfiguration.class);
        doReturn(authHeaderValue).when(authConfiguration).asAuthorizationHeaderValue();
        doReturn(authConfiguration).when(connectivity).getAuthConfiguration();
        enqueueSimpleResponse();

        fetcher.fetch();

        assertEquals(authHeaderValue, webServer.takeRequest().getHeader("Authorization"));
    }

    @Test
    public void whenEtagReceived_storeIt() throws IOException {
        String theEtag = "someEtag";
        enqueueResponse(getResponse(200, "{}").setHeader("ETag", theEtag));

        fetcher.fetch();

        verify(preferences).store("central_configuration_etag", theEtag);
    }

    @Test
    public void whenPreparingRequest_andEtagIsAvailable_sendIt() throws
            IOException, InterruptedException {
        String theEtag = "someEtag";
        doReturn(theEtag).when(preferences).retrieveString("central_configuration_etag");
        enqueueSimpleResponse();

        fetcher.fetch();

        String sentEtag = webServer.takeRequest().getHeader("If-None-Match");
        assertEquals(theEtag, sentEtag);
    }

    @Test
    public void whenCacheControlReceived_returnMaxAgeTime() throws IOException {
        int headerMaxAge = 12345;
        String headerValue = "max-age=" + headerMaxAge;
        enqueueResponse(getResponse(200, "{}").setHeader("Cache-Control", headerValue));

        FetchResult result = fetcher.fetch();

        assertEquals(headerMaxAge, result.maxAgeInSeconds.intValue());
    }

    @Test
    public void whenCacheControlReceived_WithoutMaxAge_returnMaxAgeTimeAsNull() throws
            IOException {
        String headerValue = "no-cache";
        enqueueResponse(getResponse(200, "{}").setHeader("Cache-Control", headerValue));

        FetchResult result = fetcher.fetch();

        assertNull(result.maxAgeInSeconds);
    }

    @Test
    public void whenCacheControlNotReceived_returnMaxAgeTimeAsNull() throws IOException {
        enqueueSimpleResponse();

        FetchResult result = fetcher.fetch();

        assertNull(result.maxAgeInSeconds);
    }

    @Test
    public void whenConfigurationReceived_storeInProvidedFile() throws IOException {
        String body = "{\"some\":\"configValue\"}";
        enqueueResponse(getResponse(200, body));

        fetcher.fetch();

        assertEquals(body, new String(Files.readAllBytes(configurationFile.toPath())));
    }

    private void enqueueSimpleResponse() {
        enqueueResponse(getResponse(200, "{}"));
    }

    private void enqueueResponse(MockResponse response) {
        webServer.enqueue(response);
    }

    @NonNull
    private MockResponse getResponse(int code, String body) {
        return new MockResponse().setResponseCode(code).setBody(body);
    }

    @Override
    public File getConfigurationFile() {
        return configurationFile;
    }
}

File: android-sdk/src/test/java/co/elastic/apm/android/sdk/internal/features/centralconfig/fetcher/CentralConfigurationFetcherTest.kt
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher

import co.elastic.apm.android.sdk.connectivity.ConnectivityConfiguration
import co.elastic.apm.android.sdk.features.centralconfig.fetcher.CentralConfigurationFetcher
import co.elastic.apm.android.sdk.internal.services.preferences.PreferencesService
import co.elastic.apm.android.sdk.testutils.ElasticAgentRule
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import java.io.File
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class CentralConfigurationFetcherTest {
    private lateinit var preferences: PreferencesService
    private lateinit var connectivity: ConnectivityConfiguration
    private lateinit var configurationFile: File
    private lateinit var fetcher: CentralConfigurationFetcher
    private lateinit var webServer: MockWebServer

    @get:Rule
    val agentRule = ElasticAgentRule()

    @get:Rule
    val temporaryFolder: TemporaryFolder = TemporaryFolder()

    @Before
    fun setUp() {
        setUpConnectivity()
        preferences = mockk<PreferencesService>()
        every { preferences.retrieveString("central_configuration_etag") }.returns(null)
        every { preferences.store("central_configuration_etag", any<String>()) } just Runs
        every { preferences.remove("central_configuration_etag") } just Runs
        configurationFile = temporaryFolder.newFile("configFile.json")
        fetcher = CentralConfigurationFetcher(configurationFile, preferences)
        agentRule.initialize()
    }

    @After
    fun tearDown() {
        webServer.shutdown()
    }

    private fun setUpConnectivity() {
        webServer = MockWebServer()
        connectivity = mockk<ConnectivityConfiguration>()
        setConnectivityEndpoint("")
    }

    @Test
    fun `On successful response, notify config changed`() {
        enqueueSimpleResponse()

        val fetch = fetcher.fetch(connectivity)

        assertThat(fetch.configurationHasChanged).isTrue()
    }

    @Test
    fun `When config not changed response, notify the caller`() {
        enqueueResponse(getResponse(304, ""))

        val fetch = fetcher.fetch(connectivity)

        assertThat(fetch.configurationHasChanged).isFalse()
    }

    @Test
    fun `Verify request content type`() {
        enqueueSimpleResponse()

        fetcher.fetch(connectivity)

        assertThat(webServer.takeRequest().getHeader("Content-Type")).isEqualTo("application/json")
    }

    @Test
    fun `Send headers when available`() {
        val authHeaderValue = "Bearer something"
        every { connectivity.getHeaders() }.returns(mapOf("Authorization" to authHeaderValue))
        enqueueSimpleResponse()

        fetcher.fetch(connectivity)

        assertThat(webServer.takeRequest().getHeader("Authorization")).isEqualTo(authHeaderValue)
    }

    @Test
    fun `Store eTag when received`() {
        val theEtag = "someEtag"
        enqueueResponse(getResponse(200, "{}").setHeader("ETag", theEtag))

        fetcher.fetch(connectivity)

        verify { preferences.store("central_configuration_etag", theEtag) }
    }

    @Test
    fun `Send eTag when available`() {
        val theEtag = "someEtag"
        every { preferences.retrieveString("central_configuration_etag") }.returns(theEtag)
        enqueueSimpleResponse()

        fetcher.fetch(connectivity)

        val sentEtag = webServer.takeRequest().getHeader("If-None-Match")
        assertThat(sentEtag).isEqualTo(theEtag)
    }

    @Test
    fun `When max age is provided, return it`() {
        val headerMaxAge = 12345
        val headerValue = "max-age=$headerMaxAge"
        enqueueResponse(getResponse(200, "{}").setHeader("Cache-Control", headerValue))

        val result = fetcher.fetch(connectivity)

        assertThat(result.maxAgeInSeconds?.toLong()).isEqualTo(headerMaxAge.toLong())
    }

    @Test
    fun `When max age is not provided, return null`() {
        val headerValue = "no-cache"
        enqueueResponse(getResponse(200, "{}").setHeader("Cache-Control", headerValue))

        val result = fetcher.fetch(connectivity)

        assertNull(result.maxAgeInSeconds)
    }

    @Test
    fun `When no cache control is received, return null max age`() {
        enqueueSimpleResponse()

        val result = fetcher.fetch(connectivity)

        assertNull(result.maxAgeInSeconds)
    }

    @Test
    fun `Store received config in provided file`() {
        val body = "{\"some\":\"configValue\"}"
        enqueueResponse(getResponse(200, body))

        fetcher.fetch(connectivity)

        assertThat(configurationFile.readText()).isEqualTo(body)
    }

    private fun enqueueSimpleResponse() {
        enqueueResponse(getResponse(200, "{}"))
    }

    private fun enqueueResponse(response: MockResponse) {
        webServer.enqueue(response)
    }

    private fun getResponse(code: Int, body: String): MockResponse {
        return MockResponse().setResponseCode(code).setBody(body)
    }

    private fun setConnectivityEndpoint(path: String) {
        every { connectivity.getUrl() }.returns("http://" + webServer.hostName + ":" + webServer.port + path)
        every { connectivity.getHeaders() }.returns(emptyMap())
    }
}